backend/app/models/coupon.rb

```diff
@@ -4,19 +4,35 @@ class Coupon < ApplicationRecord
   NUM_ALPHANUMERIC_CHARS_IN_TOKEN = 6
 
   belongs_to :campaign
-  belongs_to :campaign_charity, optional: true
-  has_one :secondary_donation, required: false, dependent: :nullify
+  belongs_to :redemption, optional: true
 
+  delegate :secondary_donation, to: :redemption, allow_nil: true
+  delegate :campaign_charity, to: :redemption, allow_nil: true
+
+  validates :expires_at, presence: true
   validates :url_token, presence: true, uniqueness: true
   validates :denomination, presence: true, numericality: { only_integer: true, greater_than: 0 }
-  validates_associated :secondary_donation
+  validate :coupon_expires_before_campaign_ends
 
   def redeemed?
-    campaign_charity_id.present?
+    redemption_id.present?
+  end
+
+  def expired?
+    expires_at < DateTime.current
```

Comment:
When creating we use Date.current and here we use DateTime.current. Is that okay?